### DISCUSSION TITLE
average or generate lookup table from vips_profile pixels

### DISCUSSION THREAD

**Discussion by brcarp** [libvips/libvips#2005] [Q&A]
I am trying to figure out the most efficient way to apply statistical analysis to the set of pixels identified by `vips_profile`. I'm wondering if I need to do something like generate a lookup table from its results, but that might be the wrong approach.

Here's a Ruby example:

```
irb> circle = Vips::Image.black(50, 50, bands: 3)
               .draw_circle(4.0, 25, 25, 24, fill: true)
               .draw_circle(200.0, 25, 25, 20, fill: true)
=> #<Image 50x50 uchar, 3 bands, multiband>
irb> alpha = Vips::Image.black(50, 50, bands: 1)
               .draw_circle(255.0, 25, 25, 24, fill: true)
=> #<Image 50x50 uchar, 1 bands, multiband>
irb> example = circle.copy(interpretation: 'srgb')
                .add_alpha
                .extract_band(0, n: 3)
                .bandjoin(alpha)
=> #<Image 50x50 uchar, 4 bands, srgb>
irb> prof_top_left = example.profile
=> [#<Image 50x1 int, 4 bands, histogram>, #<Image 1x50 int, 4 bands, histogram>]
irb> prof_bot_right = example.flip('horizontal').flip('vertical').profile
=> [#<Image 50x1 int, 4 bands, histogram>, #<Image 1x50 int, 4 bands, histogram>]
```

That's about as far as I've gotten with it. I'm interested in, for starters, calculating `vips_avg()` for the set of non-transparent pixels identified by `prof_top_left` and `prof_bot_right`. There are multiple approaches to excluding transparent pixels: one is to use the alpha band as a mask and another is to use boolean logic to ignore values of `50` which are the cases when the profile scanned out to the max.

The aim is to calculate an average which would be `4.0` for the example above; averaging either the entire image or even just the non-transparent pixels will of course result in a much higher value. There's a basic iterative approach in which you step through `(0..example.width)` and accumulate pixles using `getpoint()` but that's obviously slow relative to more singular operations, especially on larger images with more complicated profiles which is the intended source data. I'm just not quite sure what those operations would be in this case.

If anyone has any tips on how to isolate the profiled edge pixels for averaging or other forms of statistical analysis, please share!

**Comment by jcupitt** [comment]
Hi Brian, 

Yes the things you get from profile are just images, so you can do anything. 

I find it's usually most useful to make a one band image and then find the profile of that, for example:

```
north_down, west_right = example[3].ifthenelse(example[1], 0).profile
```

To find the profile of band 1 in non-transparent areas. From the profile, you can use min / max to find the extent of the shape, or profile again on the other axis to find the position.

Other useful operations for shape analysis:

* `project` finds row and column sums, handy together with the relational operators.
* `hist_find_indexed` lets you use one image to find the histogram of another. Combine this with eg. `xyz` and you can find the position and size of all objects in an image.
* `labelregions` will label connected regions of an image. Combine with histogram and you can find and number blobs.
* `fill_nearest` will fill zero pixels with the nearest non-zero pixel, and can calculate a distance map.
* `conv` with the right mask can do template matching.

**Comment by brcarp** [comment]
Hi John, that does give me some insight into the tools available, so thanks for that. I'm still a little unclear on how to use the position information from `profile` to select a subset of pixels for analysis. I think I could potentially do this by constructing a mask to apply with `ifthenelse` but that's the step that's missing for me (if that's even the right approach).

For example if I have a two-band image (or a 4-band that I reduce to a 2-band using `bandmean`) and I want to calculate the average value of the pixels just on the edges, I'm envisioning a process like this:

1. `north_down, west_right = twoband[1].ifthenelse(twoband[0], 0).profile`
2. `flipped = twoband.flip('horizontal').flip('vertical')`
3. `south_up, east_left = flipped[1].ifthenelse(flipped[0], 0).profile`
4. {combine `north_down, west_right, south_up, east_left` to make a 0/1 mask or LUT}(?)
5. `num_values = mask.project[0].to_a.flatten.sum` (`/ 255` if it's 0/255 instead of 0/1)
6. `sum_pixels = mask.ifthenelse(twoband[0], 0).project[0].to_a.flatten.sum`
7. `edge_average = sum_pixels / num_values`

Besides not knowing how to do step no. 4, for which I'd greatly appreciate any advice, is that also a decently efficient method to achieve that result, or are there additional shortcuts that I'm missing?

Thanks again for your help!

**Comment by jcupitt** [comment]
Sorry Brian I'm having trouble visualizing this. I think I've not really grasped what you're after.

Could you upload a sample image and desired output? What's the bigger picture of the problem you're trying to solve?

**Comment by brcarp** [comment]
> Sorry Brian I'm having trouble visualizing this. I think I've not really grasped what you're after.
> 
> Could you upload a sample image and desired output? What's the bigger picture of the problem you're trying to solve?

Sorry, I will try to make the example more clear. In the comment above, the image I'm referring to as `twoband` is a simple example that I generated using this Ruby code:

```
twoband = Vips::Image.black(51, 51, bands: 1).draw_circle(200.0, 25, 25, 24, fill: true).draw_circle(4.0, 25, 25, 20, fill: true).copy(interpretation: 'b_w').add_alpha.extract_band(0, n: 1).bandjoin(Vips::Image.black(51, 51, bands: 1).draw_circle(255.0, 25, 25, 24, fill: true))
```

![twoband](https://user-images.githubusercontent.com/57408229/83556760-77edf580-a4de-11ea-990e-ca1cefbe784b.png)

That example is just for demonstration purposes, as the problem is meant to be generalized to more complex shapes on a transparent background.

The problem I'm trying to solve is to calculate an average of just the pixels on the edge of the shape, which in this example is the light gray outer ring of the circle for which the pixels each have a value of 200.0. Thus their average is also 200.0. That is distinct from average for the entire image (`twoband.avg()`) which is 114.7, or just the first band (`twoband[0].avg()`) which is 45.39, or just the non-transparent pixels which would be a similarly small value. In the example there is a distinct region in the form of a ring but there needn't be as it could be solid or a gradient or random.

It occurred to me that `profile` can be used to identify the _positions_ of the pixels at the outermost edge of the shape, but the function returns images with their positions not with their values. My thought was that I could somehow use the positions to generate a mask or some other kind of lookup table that, when applied to the original image, would mask off just the pixels at the edge such that other operations like `project` would give me the sums necessary to determine the average.

My problem is that I'm not sure how to use the positional information yielded by `profile` to obtain the _values_ of the pixels at those positions for arithmetic. Does that make sense?

**Comment by jcupitt** [comment]
You can do things like average of pixels selected by a mask pretty easily, I guess you have that.

Making a mask from the results of profile is maybe harder. I usually do things like this with flood fills. 

https://libvips.github.io/libvips/API/current/libvips-draw.html#vips-draw-flood

In your example, you could flood from the top left to select the background, then flood again from the centre to select the black. The left over parts would be a mask for the grey area.

Could you post one of your more complex cases? Do you know your blobs are always circles, for example? Can they overlap? Are the images noisy?

**Comment by brcarp** [comment]
> Could you post one of your more complex cases? Do you know your blobs are always circles, for example? Can they overlap? Are the images noisy?

It's meant to generalize for an input set that is likely not noisy but otherwise unpredictably shaped and bound to have complex cases; imagine a block of text stylized with an outline, for example.

As far as generating a the desired result, I've had some success using `#replicate` on the images output from `#profile`, equality-compared against `.xyz`, and using that as a mask to do `#ifthenelse` followed by double calls to `#project` to get the sums of the pixel values, ultimately combined with simple division.

Here's my proof-of-concept in Ruby:

```
example = Vips::Image.black(11, 11, bands: 1)
            .draw_circle(200.0, 5, 5, 4, fill: true)
            .draw_circle(4.0, 5, 5, 2, fill: true)
            .copy(interpretation: 'b_w')
            .bandjoin(Vips::Image.black(11, 11, bands: 1)
              .draw_circle(255.0, 5, 5, 4, fill: true))

xyz = Vips::Image.xyz(example.width, example.height)

flipped = example.flipver.fliphor

north_down, west_right = example[1].ifthenelse(example[0], 0).profile
south_up, east_left = flipped[1].ifthenelse(flipped[0], 0).profile

mask = (north_down.replicate(1,xyz.height) == xyz[1]) |
       (west_right.replicate(xyz.width,1) == xyz[0]) |
       ((south_up.replicate(1,xyz.height) == xyz[1]) |
        (east_left.replicate(xyz.width,1) == xyz[0])).flipver.fliphor

mask.ifthenelse(example[0], 0).project[0].project[1].to_a.flatten[0] * 255 /
  mask.project[0].project[1].to_a.flatten[0]
=> 200
```

I'm satisfied that this works, and should generalize although I'll confirm with additional testing, but I'm wondering if there are any shortcuts I may have missed and if there are any inefficiencies here that could be further optimized with vips tools of which I may be failing to take advantage.

Thanks for walking through this with me!
